server/src/main/java/io/crate/expression/scalar/string/ParseURIFunction.java

```diff
@@ -0,0 +1,178 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.expression.scalar.string;
+
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+
+import org.elasticsearch.common.Strings;
+
+import java.nio.charset.StandardCharsets;
+
+import io.crate.data.Input;
+import io.crate.expression.scalar.ScalarFunctionModule;
+import io.crate.metadata.NodeContext;
+import io.crate.metadata.Scalar;
+import io.crate.metadata.TransactionContext;
+import io.crate.metadata.functions.Signature;
+import io.crate.types.DataTypes;
+
+public final class ParseURIFunction extends Scalar<Object, String> {
+
+    private static final String NAME = "parse_uri";
+
+    public static void register(ScalarFunctionModule module) {
+        module.register(
+            Signature.scalar(
+                NAME,
+                DataTypes.STRING.getTypeSignature(),
+                DataTypes.UNTYPED_OBJECT.getTypeSignature()
+            ),
+            ParseURIFunction::new
+        );
+    }
+
+    private final Signature signature;
+    private final Signature boundSignature;
+
+    public ParseURIFunction(Signature signature, Signature boundSignature) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+    }
+
+    @Override
+    public Signature signature() {
+        return signature;
+    }
+
+    @Override
+    public Signature boundSignature() {
+        return boundSignature;
+    }
+
+    @Override
+    @SafeVarargs
+    public final Object evaluate(TransactionContext txnCtx, NodeContext nodeCtx, Input<String>... args) {
+        String uri = args[0].value();
+        if (uri == null) {
+            return null;
+        }
+        return parseURI(uri);
+    }
+
+    private final Object parseURI(String uriText) {
+        final Map<String, Object> uriMap = new LinkedHashMap<>();
+
+        URI uri = null;
+        URL url = null;
+
+        String scheme;
+        String hostname;
+        Integer port;
+        String path;
+        String query;
+        String fragment;
+        String userinfo;
+
+        // Unfortunately the URI class is rather strict in its processing and fails for non URI-encoded characters
```

Comment:
As `URL` defines a subset of `URI` I wonder in which cases parsing it as an URI fails but succeeds with URL. If it contains non URI-encoded characters, it's not a valid URI and this method should fail.

Reply:
> I'd rather tend to be more flexibel, as the parsed Object is indeed valid in both cases. A parse_text(<URL>) method imho should follow RFC2396

Shouldn't this method then be called `parse_url` instead? and thus also supports URL's.
Anyhow, is there really any use-case to support parsing of non-valid-encoded URL's?

Reply:
I split the functions into 2 separate ones. While URLs are URIs, URLs are more specific about how e.g. the query string can be formatted. Overall the `parse_url` functions imho has a higher priority.